File: scraper/utils/typesense.py
import typesense
import json
from dotenv import load_dotenv
import os

load_dotenv()

client = typesense.Client({
  'nodes': [{
    'host': os.getenv("TYPESENSE_HOST"),
    'port': os.getenv("TYPESENSE_PORT"),      
    'protocol': os.getenv("TYPESENSE_PROTOCOL")   
  }],
  'api_key': os.getenv("TYPESENSE_API_KEY"),
  'connection_timeout_seconds': 2
})

def add_document_to_typesense(content):
    # Assume 'url' as the field name where the content's url is stored
    search_parameters = {
        "q": content['url'], 
        "query_by": "url",
        "filter_by": "url:="+content['url'],  
        "sort_by": ""
    }

    results = client.collections['contents'].documents.search(search_parameters)

    if len(results['hits']) > 0:  
        print("Document with the same url already exists")
    else:
        try:
            client.collections['contents'].documents.create(content)
            with open('content/already_added.txt', 'a') as file:
                file.write(f"{content['url']}\n")
            print("Document added successfully")
        except Exception as e:
            print(f"Failed to add document: {e}")

File: lib/typesense.ts
import Typesense from "typesense";

const client = new Typesense.Client({
  nodes: [{
    host: process.env.NEXT_PUBLIC_TYPESENSE_HOST,
    port: parseInt(process.env.NEXT_PUBLIC_TYPESENSE_PORT, 10),
    protocol: process.env.NEXT_PUBLIC_TYPESENSE_PROTOCOL
  }],
  apiKey: process.env.TYPESENSE_API_KEY,
  connectionTimeoutSeconds: 2
});

export async function contentExists(url: string): Promise<boolean> {
  try {
    const searchParameters = {
      'q': url,
      'query_by': 'url',
      'filter_by': `url:=${url}`  // Exact match filter
    };
    const result = await client.collections('contents').documents().search(searchParameters);
    return result.found > 0;
  } catch (error) {
    return false;
  }
}

export async function addContentToTypesense(contentData: { url: string; content: string; title: string; description?: string; tags?: string[]; program?: string; cve?: string; source?: string; cwe?: string }): Promise<boolean> {
  const exists = await contentExists(contentData.url);
  if (!exists) {
    try {
      await client.collections('contents').documents().create(contentData);
      return true;
    } catch (error) {
      return false;
    }
  }
  return false;
}

export async function getTotalContentsCount(): Promise<number> {
  try {
    const collection = await client.collections('contents').retrieve();
    return collection.num_documents;
  } catch (error) {
    return 0;
  }
}


